File: Lista01/Java/ex15.java
// Classe de exceção personalizada
class SaldoInsuficienteException extends Exception {
    public SaldoInsuficienteException(String mensagem) {
        super(mensagem);
    }
}

// Classe ContaBancaria
class ContaBancaria {
    private double saldo;

    public ContaBancaria(double saldoInicial) {
        this.saldo = saldoInicial;
    }

    public void sacar(double valor) throws SaldoInsuficienteException {
        if (valor > saldo) {
            throw new SaldoInsuficienteException("Saldo insuficiente para realizar o saque.");
        }
        saldo -= valor;
        System.out.println("Saque de R$" + valor + " realizado com sucesso.");
    }

    public double getSaldo() {
        return saldo;
    }
}

// Classe principal para testar
public class ex15 {
    public static void main(String[] args) {
        ContaBancaria conta = new ContaBancaria(100.0);

        try {
            conta.sacar(150.0);
        } catch (SaldoInsuficienteException e) {
            System.out.println(e.getMessage()); // Saldo insuficiente para realizar o saque.
        }

        System.out.println("Saldo atual: R$" + conta.getSaldo()); // Saldo atual: R$100.0
    }
}

File: Lista01/PY/ex15.py
class SaldoInsuficienteException(Exception):
    """Exceção personalizada para saldo insuficiente."""
    def __init__(self, mensagem="Saldo insuficiente para o saque"):
        self.mensagem = mensagem
        super().__init__(self.mensagem)

class ContaBancaria:
    def __init__(self, saldo):
        self.saldo = saldo

    def sacar(self, valor):
        if valor > self.saldo:
            raise SaldoInsuficienteException()
        self.saldo -= valor

# Exemplo de uso
def main():
    conta = ContaBancaria(100.0)
    
    try:
        conta.sacar(150.0)
    except SaldoInsuficienteException as e:
        print(e)

if __name__ == "__main__":
    main()
